### PR TITLE
Implementing Feature Switch for CFNetworkHandler and NSUrlSessionHandler

### PR DESCRIPTION
**Pull request by Agnik7**
This PR is with respect to Issue #14298 . 

Changes made:
- Added the `UseCFNetworkHandler` and `UseNSUrlSessionHandler` properties.
- Computed the values for the aforementioned properties.
- Added two new RuntimeHostConfigurationOption entries.
- Updated the `RuntimeOptions.cs` to use the two new Runtime properties.

### CODE REVIEW

**Comment by Agnik7** [comment]
@rolfbjarne I have made the changes you requested. Kindly review.

**Comment by Agnik7** [comment]
> The changes don't compile:
> 
> ```
>          /Users/builder/azdo/_work/2/s/xamarin-macios/src/ObjCRuntime/RuntimeOptions.cs(41,16): error CS1061: 'RuntimeOptions' does not contain a definition for 'UseCFNetworkHandler' and no accessible extension method 'UseCFNetworkHandler' accepting a first argument of type 'RuntimeOptions' could be found (are you missing a using directive or an assembly reference?) [/Users/builder/azdo/_work/2/s/xamarin-macios/tools/dotnet-linker/dotnet-linker.csproj]
>          /Users/builder/azdo/_work/2/s/xamarin-macios/src/ObjCRuntime/RuntimeOptions.cs(41,38): error CS0103: The name '_IsCFNetworkHandlerFeature' does not exist in the current context [/Users/builder/azdo/_work/2/s/xamarin-macios/tools/dotnet-linker/dotnet-linker.csproj]
>          /Users/builder/azdo/_work/2/s/xamarin-macios/src/ObjCRuntime/RuntimeOptions.cs(42,16): error CS1061: 'RuntimeOptions' does not contain a definition for 'UseNSUrlSessionHandler' and no accessible extension method 'UseNSUrlSessionHandler' accepting a first argument of type 'RuntimeOptions' could be found (are you missing a using directive or an assembly reference?) [/Users/builder/azdo/_work/2/s/xamarin-macios/tools/dotnet-linker/dotnet-linker.csproj]
>          /Users/builder/azdo/_work/2/s/xamarin-macios/src/ObjCRuntime/RuntimeOptions.cs(42,41): error CS0103: The name '_IsNSUrlSessionHandlerFeature' does not exist in the current context [/Users/builder/azdo/_work/2/s/xamarin-macios/tools/dotnet-linker/dotnet-linker.csproj]
>          /Users/builder/azdo/_work/2/s/xamarin-macios/src/ObjCRuntime/RuntimeOptions.cs(108,9): error CS0103: The name 'UseCFNetworkHandler' does not exist in the current context [/Users/builder/azdo/_work/2/s/xamarin-macios/tools/dotnet-linker/dotnet-linker.csproj]
>          /Users/builder/azdo/_work/2/s/xamarin-macios/src/ObjCRuntime/RuntimeOptions.cs(109,23): error CS0246: The type or namespace name 'CFNetworkHandler' could not be found (are you missing a using directive or an assembly reference?) [/Users/builder/azdo/_work/2/s/xamarin-macios/tools/dotnet-linker/dotnet-linker.csproj]
>          /Users/builder/azdo/_work/2/s/xamarin-macios/src/ObjCRuntime/RuntimeOptions.cs(112,16): error CS0103: The name 'handler_name' does not exist in the current context [/Users/builder/azdo/_work/2/s/xamarin-macios/tools/dotnet-linker/dotnet-linker.csproj]
>          /Users/builder/azdo/_work/2/s/xamarin-macios/src/ObjCRuntime/RuntimeOptions.cs(112,44): error CS0103: The name 'handler_name' does not exist in the current context [/Users/builder/azdo/_work/2/s/xamarin-macios/tools/dotnet-linker/dotnet-linker.csproj]
>          /Users/builder/azdo/_work/2/s/xamarin-macios/src/ObjCRuntime/RuntimeOptions.cs(113,16): error CS0103: The name 'Runtime' does not exist in the current context [/Users/builder/azdo/_work/2/s/xamarin-macios/tools/dotnet-linker/dotnet-linker.csproj]
>          /Users/builder/azdo/_work/2/s/xamarin-macios/src/ObjCRuntime/RuntimeOptions.cs(113,34): error CS0103: The name 'handler_name' does not exist in the current context [/Users/builder/azdo/_work/2/s/xamarin-macios/tools/dotnet-linker/dotnet-linker.csproj]
>          /Users/builder/azdo/_work/2/s/xamarin-macios/src/ObjCRuntime/RuntimeOptions.cs(114,23): error CS0246: The type or namespace name 'NSUrlSessionHandler' could not be found (are you missing a using directive or an assembly reference?) [/Users/builder/azdo/_work/2/s/xamarin-macios/tools/dotnet-linker/dotnet-linker.csproj]
> ```
> 
> did you try to [build locally](https://github.com/xamarin/xamarin-macios/wiki/Build-&-Run)?

@rolfbjarne  I wasn't able to build it locally since I have a Windows system, so I just wrote the code. Is it possible for you to build locally and suggest the required changes?

**Comment by rolfbjarne** [comment]
> @rolfbjarne I wasn't able to build it locally since I have a Windows system, so I just wrote the code. Is it possible for you to build locally and suggest the required changes?

At that point I'll end up doing more work than if I were to do it myself...

To work on this project you really need a Mac to be able to build and test your changes.

**Comment by Agnik7** [comment]
So....what should I do now?? Do you have any ideas?? Really sorry for the inconvinience caused

**Comment by dalexsoto** [comment]
@Agnik7 Thank you for your contribution, we are a bit busy right now but we will try to get back to you as soon as possible, as @rolfbjarne mentioned you do need a Mac in order to build this project and do appreciate the intent so give us some time in order to move forward.

**Comment by Agnik7** [comment]
Ok

File: src/ObjCRuntime/Runtime.cs [hunk 1]
```diff
@@ -49,6 +49,8 @@ public partial class Runtime {
 		static object lock_obj;
 		static IntPtr NSObjectClass;
 		static bool initialized;
+		static bool useCFNetworkHandler = false;
+		static bool useNSUrlSessionHandler = false;
```
Comment (rolfbjarne) [inline src/ObjCRuntime/Runtime.cs:53]: The assignment to the default value is unnecessary, except that the default is to use NSUrlSessionHandler:

```suggestion
		static bool useCFNetworkHandler;
		static bool useNSUrlSessionHandler = true;
```

**Review by rolfbjarne — CHANGES_REQUESTED** [review]
It would be nice to have some numbers to see if the linker really is able to change the app size depending on this behavior.

This can probably be done by editing the size-comparison project here:

https://github.com/xamarin/xamarin-macios/blob/main/tests/dotnet/size-comparison/MySingleView/dotnet/MySingleView.csproj

And check if there's any app size difference between:

1. No change
2. Setting `UseNativeHttpHandler=false`:

```xml
<PropertyGroup>
    <UseNativeHttpHandler>false</UseNativeHttpHandler>
</PropertyGroup>
```

3. Using NSUrlSessionHandler:

```xml
<PropertyGroup>
    <MtouchHttpClientHandler>NSUrlSessionHandler</MtouchHttpClientHandler>
</PropertyGroup>
```

3. Using CFNetworkHandler:

```xml
<PropertyGroup>
    <MtouchHttpClientHandler>CFNetworkHandler</MtouchHttpClientHandler>
</PropertyGroup>
```

Then running `make compare` in the tests/ subdirectory for each case should result in a comparison:

https://github.com/xamarin/xamarin-macios/blob/5f3c312b6ce6eb3cc0011c152bc528de19c806c9/tests/dotnet/Makefile#L17-L18

**Pull request closed without merging**